### PR TITLE
Use updated attributes when saving tokenizers

### LINKED ISSUE #20395: some tokenizer(s) don't save the updated attributes
### System Info

transformers version: 4.25.0.dev0
Torch version: 1.13.0+cpu
Cuda available: False
Cuda version: None
CuDNN version: None
Number of GPUs available: 0

### Description

For `GPT2Tokenizer(Fast)`, Set `tokenizer.model_max_length` to `128` (originally `1024`), save it then reload, will give `tokenizer.model_max_length` being `1024`.

### Reproduction

```python
from transformers import GPT2Tokenizer, GPT2TokenizerFast

tokenizer = GPT2TokenizerFast.from_pretrained("gpt2")
print(tokenizer.model_max_length)

tokenizer.model_max_length = 128
print(tokenizer.model_max_length)

tokenizer.save_pretrained("my-gpt2")
tokenizer_loaded = GPT2TokenizerFast.from_pretrained("my-gpt2")
print(tokenizer_loaded.model_max_length)
```

The output is

```bash
1024
128
1024

```

### Expected behavior

`tokenizer_loaded.model_max_length` should be `128` in the above example. In general, the updated attribute(s) should be saved.

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Use updated attributes when saving tokenizers.

Fix #20395

### CODE REVIEW

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -2082,6 +2082,14 @@ def save_pretrained(
         )
 
         tokenizer_config = copy.deepcopy(self.init_kwargs)
+
+        # TODO: Ensure the modified attributes (those are also in the __init__ kwargs) will give identical tokenizers
+        # target_keys = self.init_kwargs.keys()
+        target_keys = ["model_max_length"]
```
Comment (ydshieh) [inline src/transformers/tokenization_utils_base.py:2088]: I don't want to touch other attributes at this moment.
(although CircleCI only has 3 or 4  failures from T5 if I use `self.init_kwargs.keys()`)

**Review by sgugger — APPROVED** [review]
Fine by me, but I would like a more generic fix as proposed in the comments :-)

**Comment by ydshieh** [comment]
For the record: running slow tests with the generic fix:

2 are (By-)T5 issues, other 3 from Bert/RocBert/NLLB

```bash
=FAILED tests/models/bert/test_tokenization_bert_tf.py::BertTokenizationTest::test_saved_model - ValueError: The two structures don't have the same nested structure.
FAILED tests/models/byt5/test_tokenization_byt5.py::ByT5TokenizationTest::test_added_token_serializable - ValueError: Both extra_ids (125) and additional_special_tokens (['new_token']) are provided to ByT5Tokenizer. In this case the additional_special_tokens must include the extra_ids tokens
FAILED tests/models/nllb/test_tokenization_nllb.py::NllbTokenizationTest::test_save_pretrained - ValueError: Non-consecutive added token 'ar_AR' found. Should have index 1229 but has index 1204 in saved vocabulary.
FAILED tests/models/roc_bert/test_tokenization_roc_bert.py::BertTokenizationTest::test_sequence_builders - assert [1, 5, 6, 2] == [101, 5, 6, 102]
FAILED tests/models/t5/test_tokenization_t5.py::T5TokenizationTest::test_added_token_serializable - ValueError: Both extra_ids (100) and additional_special_tokens (['new_token']) are provided to T5Tokenizer. In this case the additional_special_tokens must include the extra_ids tokens

```

**Comment by ydshieh** [comment]
Will try to make tokenization code better, but need to merge in order to unblock tiny model creation (so for pipeline testing and ONNX testing)

**Pull request merged**